File: gtci/01-two-pointers/python3/4-triplet-sum-to-zero.py
class Solution:
    def searchTriplets(self, arr):
        triplets = []

        arr.sort()

        for idx, num in enumerate(arr):
            if idx > 0 and num == arr[idx - 1]:
                continue

            left = idx + 1
            right = len(arr) - 1

            while left < right:
                total_sum = num + arr[left] + arr[right]

                if total_sum < 0:
                    left += 1

                elif total_sum > 0:
                    right -= 1
                else:
                    triplets.append([num, arr[left], arr[right]])

                    left += 1
                    while left < right and arr[left - 1] == arr[left]:
                        left += 1

                    right -= 1
                    while left < right and arr[right + 1] == arr[right]:
                        right -= 1

        return triplets


if __name__ == "__main__":
    sol = Solution()

    arr = [-2, -1, 0, 1, 2]
    result = sol.searchTriplets(arr)
    assert sorted(result) == sorted([[-2, 0, 2], [-1, 0, 1]])

    arr2 = [-1, 0, 1, 2, -1, -4]
    result2 = sol.searchTriplets(arr2)
    assert sorted(result2) == sorted([[-1, -1, 2], [-1, 0, 1]])

    arr3 = [-3, -2, -1]
    result3 = sol.searchTriplets(arr3)
    assert sorted(result3) == sorted([])

    print("All test cases passed.")

File: gtci/01-two-pointers/cpp/4-triplet-sum-to-zero.cpp
#include <algorithm>
#include <cassert>
#include <iostream>
#include <vector>

using namespace std;

class Solution {
   public:
    static vector<vector<int>> searchTriplets(vector<int>& arr) {
        vector<vector<int>> triplets;

        sort(arr.begin(), arr.end());

        for (int i = 0; i < arr.size(); i++) {
            const int firstNum = arr[i];

            if (i > 0 && arr[i - 1] == firstNum) continue;

            vector<int>::const_iterator leftItr = arr.cbegin() + i + 1;
            vector<int>::const_iterator rightItr = arr.cend() - 1;

            while (leftItr < rightItr) {
                const int total_sum = firstNum + *leftItr + *rightItr;

                if (total_sum < 0) {
                    leftItr++;
                } else if (total_sum > 0) {
                    rightItr--;
                } else {
                    const vector<int> triplet{firstNum, *leftItr, *rightItr};
                    triplets.push_back(triplet);

                    leftItr++;
                    while (leftItr < rightItr && *(leftItr - 1) == *leftItr)
                        leftItr++;

                    rightItr--;
                    while (leftItr < rightItr && *(rightItr + 1) == *rightItr)
                        rightItr--;
                }
            }
        }

        return triplets;
    }
};

bool tripletVectorsEqual(const std::vector<std::vector<int>>& a,
                         const std::vector<std::vector<int>>& b) {
    if (a.size() != b.size()) return false;
    auto sortedA = a, sortedB = b;
    for (auto& v : sortedA) std::sort(v.begin(), v.end());
    for (auto& v : sortedB) std::sort(v.begin(), v.end());
    std::sort(sortedA.begin(), sortedA.end());
    std::sort(sortedB.begin(), sortedB.end());
    return sortedA == sortedB;
}

int main() {
    Solution sol;
    std::vector<int> arr1 = {-2, -1, 0, 1, 2};
    auto result1 = sol.searchTriplets(arr1);
    std::vector<std::vector<int>> expected1 = {{-2, 0, 2}, {-1, 0, 1}};
    assert(tripletVectorsEqual(result1, expected1));

    std::vector<int> arr2 = {-1, 0, 1, 2, -1, -4};
    auto result2 = sol.searchTriplets(arr2);
    std::vector<std::vector<int>> expected2 = {{-1, -1, 2}, {-1, 0, 1}};
    assert(tripletVectorsEqual(result2, expected2));

    std::cout << "All test cases passed." << std::endl;
    return 0;
}
